tests/jballerina-unit-test/src/test/resources/test-src/jvm/types.bal

```diff
@@ -157,6 +157,31 @@ function tupleTest() returns int {
    return 10;
 }
 
+public function testRestType() {
+    [int...] x = [1, 2];
+    any y = x;
+    assertEquality(y is string[], false);
+    assertEquality(y is [int], true);
```

Comment:
Not due to the changes you introduced, but this should also fail. :( Seems like there are issues in the `checkIsTupleType` logic too. 

The `is` check looks for the belongs-to relationship

> a value belongs to a type if it looks like the type, and it will necessarily continue to look like the type no matter how the value is mutated.

Should we fix that check in this PR too?

Reply:
so only this needs to be true? `y is [int...]`

Reply:
And `y is int[]`

Reply:
both `y is int[]` and `y is int[0]` return false now. Are rest types considered arrays as well as tuples?

Reply:
`y is int[]` should be true, `y is int[0]` should be false.

Arrays and tuples both represent lists. The difference being in an array all the members are of the same type whereas for tuples they may be of different types. Arrays and tuples can be used interchangeably given that they fulfill "belongs-to" criteria.  

The types have to match and when it comes to the size,

`int[2]` is the same as `[int, int]` - list of two int members

`int[]` is the same as `[int...]` - list of zero or more int members

`int[2]` is assignable to `[int, int, int...]` because `[int, int, int...]` represents a list of at least two int members, and `int[2]` is always that.

But you can't assign a `[int, int, int...]` to `int[2]` since it may have more than 2 members which will violate the inherent type of `int[2]`.

Reply:
Shall I create a different issue to fix this and remove the above test case?

Reply:
Okay I guess, since it's a different path. Shall we create one with high priority though, since this would lead to unexpected errors?

Reply:
Created issue https://github.com/ballerina-platform/ballerina-lang/issues/23676 . Please verify if it captures the requirements 🙂

Reply:
LGTM, thanks.